### PR TITLE
Improve test coverage and more intuitive Tags

### PR DESCRIPTION
**Pull request by polyatail**
This PR addresses the following issues:

- #168 Can now use `onecodex.Samples.where(tags='isolate')` style syntax to return a `SampleCollection` of samples matching a certain tag.
- #171 Improve test coverage and depth for viz, upload, and others

### CODE REVIEW

File: onecodex/models/sample.py [hunk 1]
```diff
@@ -24,6 +24,11 @@ def where(cls, *filters, **keyword_filters):
         public = keyword_filters.get('public', False)
         instances_route = 'instances' if not public else 'instances_public'
         limit = keyword_filters.get('limit', None if not public else 1000)
+        tag = keyword_filters.pop('tag', None)
```
Comment (boydgreenfield) [inline onecodex/models/sample.py]: So, we actually support this in the API and shouldn't need to get all samples and filter in memory. Here's an example public URL:

```
https://app.onecodex.com/api/v1/samples?where={"tags": {"$contains": {"$ref": "/api/v1/tags/dbcf5b98bca54a16"}}}
```

Consequently, I think the best way to implement this is to fetch the tag by name if we have a non-UUID (and maybe have some kind of `is_uuid()` function to skip this check) and then include it in the `.where` query directly.

Note we also implement a custom `$containsany` and `$containsall` selector in our backend that allows passing an array of tags and doing more complex intersects. We should look at the performance of that query code before leaning on it too heavily, but those may be useful too.

File: onecodex/models/sample.py [hunk 2]
```diff
@@ -75,6 +80,23 @@ def where(cls, *filters, **keyword_filters):
...
+                    if tag in s.tags:
+                        samples_to_keep.append(s)
+                except HTTPError:
+                    # this is here for some strange error where a sample has a tag that doesn't exist?
+                    pass
+
+                if len(samples_to_keep) == limit:
+                    break
+
+            samples = samples_to_keep
```
Comment (boydgreenfield) [inline onecodex/models/sample.py]: See comment above re: removing and simplifying this logic (plus shipping less data!).

File: tests/test_api_models.py [hunk 3]
```diff
@@ -81,6 +89,13 @@ def test_resourcelist(ocx, api_data):
         tags1.append(sample)
     assert 'object of type' in str(e.value)
 
+    if sys.version_info.major < 3:
+        with pytest.raises(AttributeError):
```
Comment (boydgreenfield) [inline tests/test_api_models.py]: Just out of curiosity, why does this raise? Probably worth a comment for future us.

Comment (polyatail) [reply tests/test_api_models.py]: Apparently lists in Python 2 do not have a clear() method

File: onecodex/models/sample.py (hunk 1 above)
Comment (polyatail) [reply onecodex/models/sample.py]: OK, this was a lot simpler. Changed as requested.

File: onecodex/models/sample.py [hunk 4]
```diff
@@ -25,6 +25,40 @@ def where(cls, *filters, **keyword_filters):
...
+            tags = [tags]
+
+        new_tags = []
+
+        for t in tags:
+            if isinstance(t, Tags):
+                new_tags.append(t)
+            else:
+                # is it a uuid?
+                if len(t) == 16:
```
Comment (boydgreenfield) [inline onecodex/models/sample.py]: Let's write an `is_ocx_id` helper function that does this plus and alphabet check. Something like:

```
assert set(x for x in name).issubset(string.hexdigits)
```

**Pull request merged**